src/oss-characteristics/CharacteristicTool.cs

```diff
@@ -100,33 +103,33 @@ public CharacteristicTool()
         /// </summary>
         /// <param name="purl">The package-url of the package to analyze.</param>
         /// <returns>List of tags identified</returns>
-        public async Task<Dictionary<string, AnalyzeResult>> AnalyzePackage(PackageURL purl)
+        public async Task<Dictionary<string, AnalyzeResult>> AnalyzePackage(PackageURL purl, string targetDirectoryName)
         {
             Logger.Trace("AnalyzePackage({0})", purl.ToString());
             
             var analysisResults = new Dictionary<string, AnalyzeResult>();
+            List<string> directoryNames = new List<string>();
 
-            string targetDirectoryName = null;
-            while (targetDirectoryName == null || Directory.Exists(targetDirectoryName))
+            var downloadTool = new DownloadTool();
+            // ensure that the cache directory has the required package, download it otherwise
+            bool cached = !string.IsNullOrEmpty(targetDirectoryName);
+            if (cached)
             {
-                targetDirectoryName = Path.Combine(Path.GetTempPath(), Path.GetRandomFileName());
+                directoryNames.AddRange(await downloadTool.EnsureDownloadExists(purl, targetDirectoryName));
             }
-
-            Logger.Trace("Creating directory [{0}]", targetDirectoryName);
-            Directory.CreateDirectory(targetDirectoryName);
-
-            var downloadTool = new DownloadTool();
-            var directoryNames = await downloadTool.Download(purl, targetDirectoryName);
```

Comment:
If you are not cached it looks like you no longer ever try to download

Reply:
fixed